airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/typing_deduping/SnowflakeDestinationHandler.java

```diff
@@ -30,60 +52,74 @@ public class SnowflakeDestinationHandler implements DestinationHandler<Snowflake
   private final JdbcDatabase database;
 
   public SnowflakeDestinationHandler(final String databaseName, final JdbcDatabase database) {
+    super(databaseName, database);
     this.databaseName = databaseName;
     this.database = database;
   }
 
-  @Override
-  public Optional<SnowflakeTableDefinition> findExistingTable(final StreamId id) throws SQLException {
-    // The obvious database.getMetaData().getColumns() solution doesn't work, because JDBC translates
-    // VARIANT as VARCHAR
-    final LinkedHashMap<String, SnowflakeColumnDefinition> columns = database.queryJsons(
-        """
-        SELECT column_name, data_type, is_nullable
-        FROM information_schema.columns
-        WHERE table_catalog = ?
-          AND table_schema = ?
-          AND table_name = ?
-        ORDER BY ordinal_position;
-        """,
-        databaseName.toUpperCase(),
-        id.finalNamespace().toUpperCase(),
-        id.finalName().toUpperCase()).stream()
-        .collect(LinkedHashMap::new,
-            (map, row) -> map.put(
-                row.get("COLUMN_NAME").asText(),
-                new SnowflakeColumnDefinition(row.get("DATA_TYPE").asText(), fromSnowflakeBoolean(row.get("IS_NULLABLE").asText()))),
-            LinkedHashMap::putAll);
-    if (columns.isEmpty()) {
-      return Optional.empty();
-    } else {
-      return Optional.of(new SnowflakeTableDefinition(columns));
+  public static LinkedHashMap<String, LinkedHashMap<String, TableDefinition>> findExistingTables(final JdbcDatabase database,
+                                                                                                 final String databaseName,
+                                                                                                 final List<StreamId> streamIds)
+      throws SQLException {
+    final LinkedHashMap<String, LinkedHashMap<String, TableDefinition>> existingTables = new LinkedHashMap<>();
+    final String paramHolder = String.join(",", Collections.nCopies(streamIds.size(), "?"));
+    // convert list stream to array
+    final String[] namespaces = streamIds.stream().map(StreamId::finalNamespace).toArray(String[]::new);
+    final String[] names = streamIds.stream().map(StreamId::finalName).toArray(String[]::new);
+    final String query = """
+                         SELECT table_schema, table_name, column_name, data_type, is_nullable
+                         FROM information_schema.columns
+                         WHERE table_catalog = ?
+                           AND table_schema IN (%s)
+                           AND table_name IN (%s)
+                         ORDER BY table_schema, table_name, ordinal_position;
```

Comment:
mega nit... if a user has tables:
* schema1.table1
* schema1.table2
* schema2.table2

and streams:
* schema1.table1
* schema2.table2

this query will fetch the `schema1.table2` information, even though it's not in the sync. I think you need to generate a spicy where clause to avoid that?
```
where table_catalog = ? AND (
  (table_schema = %s AND table_name = %s)
  OR (table_schema = %s AND table_name = %s)
  ...
)
```

also, just curious - why not jooqify this? I'm doing a similar thing in https://github.com/airbytehq/airbyte/pull/35303/files#diff-4057445d54a364007b364e22754d4f0b01da759cfc94262fe0c9bd2b1f218a9e / https://github.com/airbytehq/airbyte/blob/a41a52f9c11249531ff8abde5e287773429111eb/airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/java/io/airbyte/cdk/integrations/destination/jdbc/typing_deduping/JdbcDestinationHandler.java#L320-L326

Reply:
or just skip irrelevant `result` objects in your for-loop :shrug: that's probably easier, actually

Reply:
yeah. Thought about the unnecessary objects, the querying party would've have to lookup by `map.get(schema).get(tableName)`, hoping as long as the mapping order is correct it'll ignore useless entries.

Reply:
Reg Jooq. No reason, didn't venture into on this one.

Reply:
ah derp didn't notice this is in findExistingTables + we filter down to relevant streams in gatherInitialState. makes sense!